src/cirq_iqm/iqm_sampler.py

```diff
@@ -140,14 +153,27 @@ def run_iqm_batch(self, programs: list[cirq.Circuit], repetitions: int = 1) -> l
         for program in programs:
             self._device.validate_circuit(program)
 
-        measurements = self._send_circuits(programs, repetitions=repetitions)
-        return [study.ResultDict(measurements=meas) for meas in measurements]
+        circuit_results = self._send_circuits(
+            programs,
+            repetitions=repetitions,
+        )
+        results = [
+            IQMResult(
+                params=cirq.ParamResolver(),
+                measurements=result.measurements,
+                records=result.records,
+                metadata=result.metadata,
+            )
+            for result in circuit_results
+        ]
+
+        return results
 
     def _send_circuits(
         self,
         circuits: list[cirq.Circuit],
         repetitions: int = 1,
-    ) -> list[dict[str, np.ndarray]]:
+    ) -> List[IQMResult]:
```

Comment:
Two comments:
* where possible we use lowercase types because they do not need to be imported: here `list` vs `List`
* I would just return a `list[tuple[dict[str, np.ndarray],ResultMetadata]]` here because otherwise you are constructing `IQMResult` twice.

Reply:
Thanks for the comments. I have made `_create_iqm_result` a static method now. And where possible I am using lowercase types rather than their uppercase equivalents. Note that this means I needed to cast certain variables (e.g `measurements` and `records` to `dict` in order to satisfy mypy. For reference the test error was 

```bash
163: error: Argument "measurements" to "IQMResult" has incompatible type "Mapping[str, ndarray[Any, Any]]"; expected "Optional[Dict[str, ndarray[Any, Any]]]"
164: error: Argument "records" to "IQMResult" has incompatible type "Mapping[str, ndarray[Any, Any]]"; expected "Optional[Dict[str, ndarray[Any, Any]]]"
231: error: Argument "measurements" to "IQMResult" has incompatible type "Mapping[str, ndarray[Any, Any]]"; expected "Optional[Dict[str, ndarray[Any, Any]]]"
232: error: Argument "records" to "IQMResult" has incompatible type "Mapping[str, ndarray[Any, Any]]"; expected "Optional[Dict[str, ndarray[Any, Any]]]"
```

I would argue that using `Mapping` from `typing` would make this simpler. 

Regarding the second comment, I attempted to implement this but the types got very messy and passing the tests was difficult. I'll have to either think more on it or ask for help from you.  Using `list[tuple[dict[str, np.ndarray],ResultMetadata]]` as the return type hint for `_send_circuits` required using in in the subsequent function as well such as `run_sweep` and `run_iqm_batch`, however as IQMResults is defined this gave many type errors

```bash
124: error: "Tuple[Dict[str, ndarray[Any, Any]], ResultMetadata]" has no attribute "measurements"
125: error: "Tuple[Dict[str, ndarray[Any, Any]], ResultMetadata]" has no attribute "records"
126: error: "Tuple[Dict[str, ndarray[Any, Any]], ResultMetadata]" has no attribute "metadata"
163: error: "Tuple[Dict[str, ndarray[Any, Any]], ResultMetadata]" has no attribute "measurements"
164: error: "Tuple[Dict[str, ndarray[Any, Any]], ResultMetadata]" has no attribute "records"
165: error: "Tuple[Dict[str, ndarray[Any, Any]], ResultMetadata]" has no attribute "metadata"
```

And additionally some mypy errors got introduced in `optimizers.py` and not sure how?

```bash
199: error: Need type annotation for "moment"
```

Reply:
Hi @JMuff22 ,
Looks good. Some of the errors you are observing seem to come from an update of third party dependencies. I have drafted a fix [here](https://github.com/iqm-finland/cirq-on-iqm/pull/115). Will check this internally and then we can continue on this.